File: app/src/main/java/bootpay/co/kr/samplepayment/BootpayWebView.java
package bootpay.co.kr.samplepayment;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageManager;
import android.net.ConnectivityManager;
import android.net.Uri;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import android.provider.Settings;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.JavascriptInterface;
import android.webkit.JsResult;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import java.net.URISyntaxException;
import java.util.List;
import java.util.Locale;
import java.util.StringJoiner;

import bootpay.co.kr.samplepayment.model.Item;
import bootpay.co.kr.samplepayment.model.Request;

public final class BootpayWebView extends WebView {

    private static final String BOOTPAY = "https://dev-app.bootpay.co.kr";

    private ConnectivityManager connManager;

    private AlertDialog networkErrorDialog;

    private static final int ERROR = -2;

    private static final int CANCEL = -1;

    private static final int CONFIRM = 1;

    private static final int DONE = 2;

    private Request request = null;

    private Locale locale = Locale.getDefault();

    private Handler eventHandler = new Handler(Looper.getMainLooper()) {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            String data = msg.obj.toString();
            switch (msg.what) {
                case ERROR:
                    onErrorHandled(data);
                    break;
                case CANCEL:
                    onCancelHandled(data);
                    break;
                case CONFIRM:
                    onConfirmeHandled(data);
                    break;
                case DONE:
                    onDoneHandled(data);
                    break;
            }
        }
    };

    private EventListener listener;

    public BootpayWebView(Context context) {
        this(context, null);
    }

    public BootpayWebView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public BootpayWebView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        connManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        setFocusable(true);
        setFocusableInTouchMode(true);
        requestFocus();

        setWebViewClient(new WebViewClient() {
            @Override
            public void onPageFinished(WebView view, String url) {
                super.onPageFinished(view, url);
                setDevice();
                load(
                        request(
                                price(),
                                application_id(),
                                name(),
                                pg(),
                                method(),
                                items(),
                                test_mode(),
                                order_id()
                        ),
                        error(),
                        cancel(),
                        confirm(),
                        done() + ";"
                );
            }

            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                Intent intent = parse(url);
                if (isIntentOf(url)) {
                    if (isExistInfo(intent) || isExistPackage(intent)) start(intent);
                    else startMarket(intent);
                } else if (isMarketOf(url)) start(intent);
                return true;
            }
        });

        if (isNetworkConnected()) {
            setting(context);
            loadUrl(BOOTPAY);
        } else networkErrorDialog(context).show();
    }

    boolean back(Dialog dialog) {
        if (canGoBack()) {
            Log.d("Back", "Can go back");
            goBack();
        } else {
            dialog.onBackPressed();
            Log.d("Back", "Can't go back");
        }
        return true;
    }

    private boolean isIntentOf(String url) {
        return url != null && url.startsWith("intent://");
    }

    private boolean isMarketOf(String url) {
        return url != null && url.startsWith("market://");
    }

    private boolean isExistInfo(Intent intent) {
        try {
            return intent != null && getContext().getPackageManager().getPackageInfo(intent.getPackage(), PackageManager.GET_ACTIVITIES) != null;
        } catch (PackageManager.NameNotFoundException e) {
            return false;
        }
    }

    private boolean isExistPackage(Intent intent) {
        return intent != null && getContext().getPackageManager().getLaunchIntentForPackage(intent.getPackage()) != null;
    }

    private Intent parse(String url) {
        try {
            return Intent.parseUri(url, Intent.URI_INTENT_SCHEME);
        } catch (URISyntaxException e) {
            return null;
        }
    }

    private void start(Intent intent) {
        getContext().startActivity(intent);
    }

    private void setDevice() {
        load("window.BootPay.setDevice('ANDROID');");
    }

    private void startMarket(Intent intent) {
        Intent market = new Intent(Intent.ACTION_VIEW);
        market.setData(Uri.parse("market://details?id=" + intent.getPackage()));
        start(market);
    }

    private String request(String... query) {
        StringBuilder builder = new StringBuilder();
        builder.append("BootPay.request({");
        for (String s : query) builder.append(s).append(",");
        builder.setLength(builder.length() - 1);
        builder.append("})");
        return builder.toString();
    }

    private String error() {
        return ".error(function(data){Android.error(JSON.stringify(data));})";
    }

    private String confirm() {
        return ".confirm(function(data){Android.confirm(JSON.stringify(data));this.transactionConfirm(data);})";
    }

    private String cancel() {
        return ".cancel(function(data){Android.cancel(JSON.stringify(data));})";
    }

    private String done() {
        return ".done(function(data){Android.done(JSON.stringify(data));})";
    }

    private String price() {
        return "price:" + request.getPrice();
    }

    private String application_id() {
        return String.format(locale, "application_id:'%s'", request.getApplication_id());
    }

    private String name() {
        return String.format(locale, "name:'%s'", request.getName());
    }

    private String pg() {
        return String.format(locale, "pg:'%s'", request.getPg());
    }

    private String method() {
        return String.format(locale, "method:'%s'", request.getMethod());
    }

    private String test_mode() {
        return String.format(locale, "test_mode:%s", request.isTest_mode());
    }

    private String items() {
        StringBuilder builder = new StringBuilder().append("items:[");
        List<Item> items = request.getItems();
        int size = items.size();
        if (size > 0) for (int i = 0; i < size; i++) {
            Item item = items.get(i);
            if (item != null) {
                builder.append("{");
                builder.append(String.format(locale, "item_name:'%s',", item.getName()));
                builder.append(String.format(locale, "qty:%d,", item.getQuantity()));
                builder.append(String.format(locale, "unique:'%s',", item.getPrimaryKey()));
                builder.append("price: ").append(item.getPrice());
                if (i != size - 1) builder.append("},");
                else builder.append("}");
            }
        }
        builder.append("]");
        return builder.toString();
    }

    private String order_id() {
        return String.format(locale, "order_id:'%s'", request.getOrderId());
    }

    void setData(Request request) {
        this.request = request;
    }

    private AlertDialog networkErrorDialog(Context context) {
        if (networkErrorDialog == null) networkErrorDialog = new AlertDialog.Builder(context)
                .setMessage("인터넷 연결이 끊어져 있습니다. 인터넷 연결 상태를 확인해주세요.")
                .setPositiveButton("설정", (d, i) -> context.startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS)))
                .setNegativeButton("확인", (d, i) -> d.dismiss())
                .create();
        return networkErrorDialog;
    }

    private boolean isNetworkConnected() {
        return connManager.getActiveNetworkInfo() != null;
    }

    private void load(String script) {
        loadUrl(String.format(locale, "javascript:(function(){%s})()", script));
    }

    private void load(String... script) {
        StringBuilder builder = new StringBuilder();
        for (String s : script) builder.append(s);
        String request = builder.toString();
        Log.d("request", request);
        load(request);
    }

    @SuppressLint("setJavaScriptEnabled")
    private void setting(Context context) {
        setWebChromeClient(new Client());
        addJavascriptInterface(new AndroidBridge(), "Android");
        CookieManager.getInstance().setAcceptCookie(true);
        if (Build.VERSION.SDK_INT >= 21)
            CookieManager.getInstance().setAcceptThirdPartyCookies(this, true);
        WebSettings s = getSettings();
        s.setJavaScriptEnabled(true);
        s.setJavaScriptCanOpenWindowsAutomatically(true);
        s.setDomStorageEnabled(true);
        s.setSupportMultipleWindows(true);
        s.setLoadsImagesAutomatically(true);
        s.setBuiltInZoomControls(true);
        s.supportZoom();
        s.setUseWideViewPort(true);
        s.setAppCacheEnabled(true);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT)
            if (0 != (context.getApplicationInfo().flags &= ApplicationInfo.FLAG_DEBUGGABLE))
                WebView.setWebContentsDebuggingEnabled(true);

    }

    void setOnResponseListener(EventListener listener) {
        this.listener = listener;
    }

    private class AndroidBridge implements IScriptFuction {

        @JavascriptInterface
        @Override
        public void error(String data) {
            eventHandler.sendMessage(Message.obtain(eventHandler, ERROR, data));
        }

        @JavascriptInterface
        @Override
        public void cancel(String data) {
            eventHandler.sendMessage(Message.obtain(eventHandler, CANCEL, data));
        }

        @JavascriptInterface
        @Override
        public void confirm(String data) {
            eventHandler.sendMessage(Message.obtain(eventHandler, CONFIRM, data));
        }

        @JavascriptInterface
        @Override
        public void done(String data) {
            eventHandler.sendMessage(Message.obtain(eventHandler, DONE, data));
        }
    }

    private void onErrorHandled(String data) {
        if (listener != null) listener.onError(data);
    }

    private void onCancelHandled(String data) {
        if (listener != null) listener.onCancel(data);
    }

    private void onConfirmeHandled(String data) {
        if (listener != null) listener.onConfirmed(data);
    }

    private void onDoneHandled(String data) {
        if (listener != null) listener.onDone(data);
    }

    private class Client extends WebChromeClient {

        @Override
        public boolean onCreateWindow(WebView view, boolean isDialog, boolean isUserGesture, Message resultMsg) {
            WebView.HitTestResult result = view.getHitTestResult();
            String url = result.getExtra();

            if (isNewWindow(url)) {
                start(view, createFrom(url));
            } else if (resultMsg != null) {
                WebView newWindow = new WebView(view.getContext());
                addView(newWindow);
                WebView.WebViewTransport tr = (WebView.WebViewTransport) resultMsg.obj;
                tr.setWebView(newWindow);
                resultMsg.sendToTarget();
            }
            return true;
        }

        @Override
        public void onCloseWindow(WebView window) {
            super.onCloseWindow(window);
            removeView(window);
            window.setVisibility(View.GONE);

        }

        @Override
        public boolean onJsAlert(WebView view, String url, String message, JsResult result) {
            AlertDialog dialog = new AlertDialog.Builder(view.getContext())
                    .setMessage(message)
                    .setCancelable(true)
                    .create();
            dialog.setCanceledOnTouchOutside(true);
            dialog.show();
            result.confirm();
            return true;
        }

        private void start(View view, Intent intent) {
            view.getContext().startActivity(intent);
        }

        private boolean isNewWindow(String url) {
            return url != null && !url.contains("___target=_blank");
        }

        private Intent createFrom(String url) {
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(Uri.parse(url));
            return intent;
        }
    }
}

File: bootpay/src/main/java/kr/co/bootpay/BootpayWebView.kt
package kr.co.bootpay

//import kr.co.bootpay.model.Trace
import android.annotation.SuppressLint
import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.pm.ApplicationInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.AttributeSet
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.webkit.*
import kr.co.bootpay.listner.EventListener
import kr.co.bootpay.model.Request
import kr.co.bootpay.pref.UserInfo
import java.net.URISyntaxException

internal class BootpayWebView @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0): WebView(context, attrs, defStyleAttr) {

    companion object {
        private const val BOOTPAY = "https://inapp.bootpay.co.kr/2.1.1/production.html"

        private const val CLOSE = -3

        private const val ERROR = -2

        private const val CANCEL = -1

        private const val READY = 0

        private const val CONFIRM = 1

        private const val DONE = 2
    }

    private var request: Request? = null

    private var isLoaded = false

    private val eventHandler = object: Handler(Looper.getMainLooper()) {
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            val data = msg.obj.toString()
            when (msg.what) {
                CLOSE   -> onCloseHandled(data)
                ERROR   -> onErrorHandled(data)
                CANCEL  -> onCancelHandled(data)
                READY   -> onReadyHandled(data)
                CONFIRM -> onConfirmHandled(data)
                DONE    -> onDoneHandled(data)
            }
        }
    }

    private var listener: EventListener? = null

    private var dialog: Dialog? = null

    init {
        isFocusable = true
        isFocusableInTouchMode = true

        requestFocus()
        setting(context)

        setWebViewClient(object: WebViewClient() {

            override fun onPageFinished(view: WebView, url: String) {
                super.onPageFinished(view, url)
                if (!isLoaded) { // recursive called under version 5.0 Lollipop OS
                    isLoaded = true
                    registerAppId()

                    setDevice()
                    setAnalyticsData(
                            UserInfo.bootpay_uuid,
                            UserInfo.bootpay_sk,
                            UserInfo.bootpay_last_time,
                            System.currentTimeMillis() - UserInfo.bootpay_last_time)

                    loadParams(
                            request(
                                    price(),
                                    applicationId(),
                                    name(),
                                    pg(),
//                                    userPhone(),
                                    agree(),
                                    method(),
                                    methods(),
                                    items(),
                                    params(),
                                    accountExpireAt(),
                                    orderId(),
                                    useOrderId(),
                                    userJson(),
                                    extraJson(),
                                    remoteLinkJson()
                            ),
                            error(),
                            cancel(),
                            ready(),
                            confirm(),
                            close(),
                            done()
                    )
                }
            }

            override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
                val intent = parse(url)
                return if (isIntent(url)) {
                    if (isExistInfo(intent) or isExistPackage(intent))
                        start(intent)
                    else
                        gotoMarket(intent)
                }
                else if (isMarket(url))
                    start(intent)
                else if (isSpecialCase(url))
                    start(intent)
                else
                    url.contains("https://testquickpay")
            }

            override fun shouldOverrideKeyEvent(view: WebView, event: KeyEvent): Boolean {
                if (event.keyCode == KeyEvent.KEYCODE_BACK) back()
                return super.shouldOverrideKeyEvent(view, event)
            }
        })

        loadUrl(BOOTPAY)
    }

    fun setDialog(dialog: Dialog): BootpayWebView {
        this.dialog = dialog
        return this
    }

    fun back(): Boolean {
        if (canGoBack()) goBack()
        else dialog?.cancel()
        return true
    }

    private fun isIntent(url: String?) = url?.matches(Regex("^intent:?\\w*://\\S+$")) ?: false

    private fun isMarket(url: String?) = url?.matches(Regex("^market://\\S+$")) ?: false

    private fun isExistInfo(intent: Intent?): Boolean {
        return try {
            intent != null && context.packageManager.getPackageInfo(intent.`package`, PackageManager.GET_ACTIVITIES) != null
        } catch (e: PackageManager.NameNotFoundException) {
            false
        }

    }

    private fun isExistPackage(intent: Intent?): Boolean =
            intent != null && context.packageManager.getLaunchIntentForPackage(intent.`package`) != null

    private fun isSpecialCase(url: String?) = url?.matches(Regex("^shinhan\\S+\$")) ?: false

    private fun parse(url: String): Intent? {
        return try {
            Intent.parseUri(url, Intent.URI_INTENT_SCHEME)
        } catch (e: URISyntaxException) {
            null
        }

    }

    private fun start(intent: Intent?): Boolean {
        intent?.let { context.startActivity(it) }
        return true
    }

    private fun setAnalyticsData(uuid: String, sk: String, sk_time: Long, time: Long) {
        load("window.BootPay.setAnalyticsData({uuid:'$uuid',sk:'$sk',sk_time:'$sk_time',time:'$time'});")
    }

    private fun setDevice() {
        load("window.BootPay.setDevice('ANDROID');")
    }

    private fun gotoMarket(intent: Intent?): Boolean {
        intent?.let { start(Intent(Intent.ACTION_VIEW).apply { data = Uri.parse("market://details?id=${it.`package`}") }) }
        return true
    }

    private fun request(vararg query: String) = "BootPay.request({${query.filter(String::isNotEmpty).joinToString()}})"

//    private fun userJson(vararg info: String) = "user_info: {${info.filter(String::isNotEmpty).joinToString()}}"
//    private fun userPhone() = request?.bootUser?.phone?.takeIf(String::isNotEmpty)?.let { "phone: '$it'" } ?: ""
    private fun userJson() = request?.bootUser?.let { "user_info: ${it.toJson()}" } ?: ""
    private fun extraJson() = request?.bootExtra?.let { "bootExtra: ${it.toJson()}" } ?: ""
    private fun remoteLinkJson() = request?.remoteLink?.let { "remte_link_info: ${it.toJson()}" } ?: ""



//    private fun bootExtra(vararg etcs: String) = "bootExtra: {${etcs.filter(String::isNotEmpty).joinToString()}}"

//    private
//    private fun userName() = request?.bootUser?.name?.takeIf(String::isNotEmpty)?.let { "username: '$it'" } ?: ""
//
//    private fun userEmail() = request?.bootUser?.email?.takeIf(String::isNotEmpty)?.let { "email: '$it'" } ?: ""
//
//    private fun userAddr() = request?.user_addr?.takeIf(String::isNotEmpty)?.let { "addr: '$it'" } ?: ""
//
//    private fun userPhone() = request?.user_phone?.takeIf(String::isNotEmpty)?.let { "phone: '$it'" } ?: ""



//    private fun extraStartAt() = request?.extra_start_at?.let { "start_at: '$it'" } ?: ""
//    private fun extraEndAt() = request?.extra_end_at?.let { "end_at: '$it'" } ?: ""
//    private fun extraExpireMonth() = request?.extra_expire_month?.let { "expire_month: $it" } ?: ""
//    private fun extraVBankResult() = request?.extra_vbank_result?.let { "vbank_result: $it" } ?: ""
//    private fun extraQuota() = request?.extra_quotas?.let { "quota: '${it.joinToString()}'" } ?: ""
//    private fun extraAppScheme() = request?.extra_app_scheme?.let { "app_scheme: '${it}://${extraAppSchemeHost()}'" } ?: ""
//    private fun extraAppSchemeHost() = request?.extra_app_scheme_host?.let { "${it}" } ?: ""


    private fun error() = ".error(function(data){Android.error(JSON.stringify(data));})"

    private fun ready() = ".ready(function(data){Android.ready(JSON.stringify(data));})"

    private fun close() = ".close(function(data){Android.close('close');})"

    private fun confirm() = ".confirm(function(data){Android.confirm(JSON.stringify(data));})"

    private fun cancel() = ".cancel(function(data){Android.cancel(JSON.stringify(data));})"

    private fun done() = ".done(function(data){Android.done(JSON.stringify(data));})"

    private fun price() = request?.price?.let { "price:$it" } ?: ""

    private fun applicationId() = request?.application_id?.let { "application_id:'$it'" } ?: ""

    private fun name() = request?.name?.let { "name:'${it.replace("\"", "'").replace("'", "\\'")}'" } ?: ""

    private fun pg() = request?.pg?.let { "pg:'$it'" } ?: ""

    private fun agree() = "show_agree_window: ${if (request?.is_show_agree == true) 1 else 0}"

    private fun method() = request?.method?.let { "method:'$it'" } ?: ""

    private fun methods() = request?.methods?.let { "methods: ${listToString(it)}"} ?:""

    private fun items() = "items:${
    request?.items?.map { "{item_name:'${it.item_name.replace("\"", "'").replace("'", "\\'")}',qty:${it.qty},unique:'${it.unique}',price:${it.price},cat1:'${it.cat1.replace("\"", "'").replace("'", "\\'")}',cat2:'${it.cat2.replace("\"", "'").replace("'", "\\'")}',cat3:'${it.cat3.replace("\"", "'").replace("'", "\\'")}'}" }
    }"

    private fun orderId() = request?.order_id?.let { "order_id:'$it'" } ?: ""

    private fun useOrderId() = request?.use_order_id?.let { "use_order_id:'$it'" } ?: "0"

    private fun listToString(list: List<String>):String {
        var buffer = ""
        for(obj in list) {
            if(buffer.length > 0) buffer += ","
            buffer += "'$obj'"
        }
        return "[$buffer]"
    }

    private fun accountExpireAt() = request?.account_expire_at?.let { "account_expire_at:'$it'" } ?: ""

    private fun params(): String {
        return if (request?.params.isNullOrEmpty())
            "params:''"
        else
            "params:JSON.parse('${request!!.params}')"
    }

    private fun registerAppId() {
        load("window.BootPay.setApplicationId('${request!!.application_id}');")
//        load("$(\"script[data-boot-app-id]\").attr(\"data-boot-app-id\", '${request!!.application_id}');")
    }

    fun transactionConfirm(data: String?) {
        load("window.BootPay.transactionConfirm(JSON.parse('${data ?: ""}'));")
    }

    fun removePaymentWindow() {
        load("window.BootPay.removePaymentWindow();")
    }

//    private fun items() = "items:${
//    request?.items?.map { "{item_name:${it.name},qty:${it.qty},unique:'${it.unique}',price:${it.price}},cat1:${it.cat1},cat2:${it.cat2},cat3:${it.cat3}," }?.dropLast(2)
//    }"


    fun setData(request: Request?): BootpayWebView {
        this.request = request
        return this
    }

    private fun load(script: String) {
        loadUrl("javascript:(function(){$script})()")
    }

    private fun loadParams(vararg script: String) {
        load("${script.joinToString("")};")
    }

    @SuppressLint("setJavaScriptEnabled")
    private fun setting(context: Context) {
        setWebChromeClient(Client())
        addJavascriptInterface(AndroidBridge(), "Android")
        CookieManager.getInstance().setAcceptCookie(true)

        settings.apply {
            if (Build.VERSION.SDK_INT >= 21) {
                mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
                CookieManager.getInstance().setAcceptCookie(true)
                CookieManager.getInstance().setAcceptThirdPartyCookies(this@BootpayWebView, true)
            }
            javaScriptEnabled = true
            javaScriptCanOpenWindowsAutomatically = true
            domStorageEnabled = true
            setSupportMultipleWindows(true)
            loadsImagesAutomatically = true
            useWideViewPort = true
            loadWithOverviewMode = true
            setAppCacheEnabled(true)
        }
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            context.applicationInfo.flags = context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE
            if (0 != (context.applicationInfo.flags))
                WebView.setWebContentsDebuggingEnabled(true)
        }
    }

    fun setOnResponseListener(listener: EventListener?) {
        this.listener = listener
    }

    private inner class AndroidBridge: IScriptFuction {
        @JavascriptInterface
        override fun close(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, CLOSE, data))
        }

        @JavascriptInterface
        override fun ready(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, READY, data))
        }

        @JavascriptInterface
        override fun error(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, ERROR, data))
        }

        @JavascriptInterface
        override fun cancel(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, CANCEL, data))
        }

        @JavascriptInterface
        override fun confirm(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, CONFIRM, data))
        }

        @JavascriptInterface
        override fun done(data: String) {
            eventHandler.sendMessage(Message.obtain(eventHandler, DONE, data))
        }
    }

    private fun onCloseHandled(data: String) {
        listener?.onClose(data)
        dialog?.dismiss()
    }

    private fun onErrorHandled(data: String) {
        listener?.onError(data)
    }

    private fun onCancelHandled(data: String) {
        listener?.onCancel(data)
    }

    private fun onReadyHandled(data: String) {
        listener?.onReady(data)
    }

    private fun onConfirmHandled(data: String) {
        listener?.onConfirm(data)
    }

    private fun onDoneHandled(data: String) {
        listener?.onDone(data)
    }

    private inner class Client: WebChromeClient() {

        override fun onCreateWindow(view: WebView, isDialog: Boolean, isUserGesture: Boolean, resultMsg: Message?): Boolean {
            val result = view.hitTestResult
            val url = result.extra

            if (isNewWindow(url)) {
                start(view, createFrom(url))
            } else if (resultMsg != null) {
                val newWindow = WebView(view.context)
                addView(newWindow)
                val tr = resultMsg.obj as WebView.WebViewTransport
                tr.webView = newWindow
                resultMsg.sendToTarget()
            }
            return true
        }

        override fun onCloseWindow(window: WebView) {
            super.onCloseWindow(window)
            removeView(window)
            window.visibility = View.GONE

        }

        override fun onJsAlert(view: WebView, url: String, message: String, result: JsResult): Boolean {
            AlertDialog.Builder(view.context)
                    .setMessage(message)
                    .setCancelable(true)
                    .create()
                    .apply {
                        setCanceledOnTouchOutside(true)
                        show()
                    }
            result.confirm()
            return true
        }

        private fun start(view: View, intent: Intent) {
            view.context.startActivity(intent)
        }

        private fun isNewWindow(url: String?): Boolean =
                url != null && !url.contains("___target=_blank")

        private fun createFrom(url: String): Intent {
            val intent = Intent(Intent.ACTION_VIEW)
            intent.data = Uri.parse(url)
            return intent
        }
    }
}
